File: CashuWallet/Core/PaymentRequestBuilder.swift
import Foundation

enum PaymentRequestBuilder {
    enum BuildError: Error {
        case invalidPubkeyHex
        case nprofileEncodeFailed
    }

    /// Build a NUT-18 creqA-encoded payment request.
    /// Wire format: "creqA" + base64url(no padding)(CBOR(payload)).
    static func build(
        id: String,
        amount: UInt64?,
        unit: String?,
        singleUse: Bool? = nil,
        mints: [String],
        description: String?,
        nostrPubkeyHex: String,
        relays: [String],
        nip: String = "17"
    ) throws -> String {
        let nprofile = try makeNprofile(pubkeyHex: nostrPubkeyHex, relays: relays)

        var transport: [(Nut18Key, Nut18Value)] = []
        transport.append((.text("t"), .text("nostr")))
        transport.append((.text("a"), .text(nprofile)))
        transport.append((.text("g"), .array([
            .array([.text("n"), .text(nip)])
        ])))

        var request: [(Nut18Key, Nut18Value)] = []
        request.append((.text("i"), .text(id)))
        if let amount, amount > 0 {
            request.append((.text("a"), .uint(amount)))
        }
        if let unit, !unit.isEmpty {
            request.append((.text("u"), .text(unit)))
        }
        if let singleUse {
            request.append((.text("s"), .bool(singleUse)))
        }
        if !mints.isEmpty {
            request.append((.text("m"), .array(mints.map { .text($0) })))
        }
        if let description, !description.isEmpty {
            request.append((.text("d"), .text(description)))
        }
        request.append((.text("t"), .array([.map(transport)])))

        let cbor = Nut18CBOR.encode(.map(request))
        return "creqA" + Base64URL.encode(cbor)
    }

    /// Encode a NIP-19 nprofile: bech32(hrp="nprofile", TLV(pubkey, relays...))
    static func makeNprofile(pubkeyHex: String, relays: [String]) throws -> String {
        guard let pubkeyBytes = Data(hex: pubkeyHex), pubkeyBytes.count == 32 else {
            throw BuildError.invalidPubkeyHex
        }
        var tlv = Data()
        // Type 0: pubkey (32 bytes)
        tlv.append(0x00)
        tlv.append(UInt8(pubkeyBytes.count))
        tlv.append(pubkeyBytes)
        // Type 1: relay URL (utf-8)
        for relay in relays {
            let bytes = Array(relay.utf8)
            guard bytes.count <= 255 else { continue }
            tlv.append(0x01)
            tlv.append(UInt8(bytes.count))
            tlv.append(contentsOf: bytes)
        }
        do {
            return try Bech32.encode(hrp: "nprofile", data: tlv)
        } catch {
            throw BuildError.nprofileEncodeFailed
        }
    }
}

// MARK: - Hex helper

extension Data {
    init?(hex: String) {
        let cleaned = hex.hasPrefix("0x") ? String(hex.dropFirst(2)) : hex
        guard cleaned.count.isMultiple(of: 2) else { return nil }
        var bytes = Data()
        bytes.reserveCapacity(cleaned.count / 2)
        var index = cleaned.startIndex
        while index < cleaned.endIndex {
            let next = cleaned.index(index, offsetBy: 2)
            guard let byte = UInt8(cleaned[index..<next], radix: 16) else { return nil }
            bytes.append(byte)
            index = next
        }
        self = bytes
    }
}

// MARK: - Minimal CBOR (RFC 8949 — definite-length subset)

enum Nut18Key {
    case text(String)
}

indirect enum Nut18Value {
    case text(String)
    case uint(UInt64)
    case bool(Bool)
    case array([Nut18Value])
    case map([(Nut18Key, Nut18Value)])
}

enum Nut18CBOR {
    static func encode(_ value: Nut18Value) -> Data {
        var out = Data()
        encodeInto(value, &out)
        return out
    }

    private static func encodeInto(_ value: Nut18Value, _ out: inout Data) {
        switch value {
        case .text(let s):
            let bytes = Array(s.utf8)
            writeHeader(majorType: 3, length: UInt64(bytes.count), into: &out)
            out.append(contentsOf: bytes)
        case .uint(let n):
            writeHeader(majorType: 0, length: n, into: &out)
        case .bool(let b):
            out.append(b ? 0xF5 : 0xF4)
        case .array(let items):
            writeHeader(majorType: 4, length: UInt64(items.count), into: &out)
            for item in items { encodeInto(item, &out) }
        case .map(let pairs):
            writeHeader(majorType: 5, length: UInt64(pairs.count), into: &out)
            for (k, v) in pairs {
                if case .text(let s) = k {
                    encodeInto(.text(s), &out)
                }
                encodeInto(v, &out)
            }
        }
    }

    private static func writeHeader(majorType: UInt8, length: UInt64, into out: inout Data) {
        let m = majorType << 5
        if length < 24 {
            out.append(m | UInt8(length))
        } else if length < 0x100 {
            out.append(m | 24)
            out.append(UInt8(length))
        } else if length < 0x10000 {
            out.append(m | 25)
            out.append(UInt8(length >> 8))
            out.append(UInt8(length & 0xFF))
        } else if length < 0x100000000 {
            out.append(m | 26)
            for shift in stride(from: 24, through: 0, by: -8) {
                out.append(UInt8((length >> shift) & 0xFF))
            }
        } else {
            out.append(m | 27)
            for shift in stride(from: 56, through: 0, by: -8) {
                out.append(UInt8((length >> shift) & 0xFF))
            }
        }
    }
}

// MARK: - Base64URL (no padding)

enum Base64URL {
    static func encode(_ data: Data) -> String {
        var s = data.base64EncodedString()
        s = s.replacingOccurrences(of: "+", with: "-")
        s = s.replacingOccurrences(of: "/", with: "_")
        while s.hasSuffix("=") { s.removeLast() }
        return s
    }

    static func decode(_ s: String) -> Data? {
        var t = s.replacingOccurrences(of: "-", with: "+")
        t = t.replacingOccurrences(of: "_", with: "/")
        while t.count % 4 != 0 { t.append("=") }
        return Data(base64Encoded: t)
    }
}

File: android/app/src/main/java/com/cashu/me/Core/PaymentRequestBuilder.kt
package com.cashu.me.Core

import java.util.Base64

object PaymentRequestBuilder {
    fun build(
        id: String,
        amount: Long?,
        unit: String?,
        singleUse: Boolean? = null,
        mints: List<String>,
        description: String?,
        nostrPubkeyHex: String,
        relays: List<String>,
        nip: String = "17",
        p2pkPubkeyHex: String? = null,
    ): String {
        val nprofile = makeNprofile(nostrPubkeyHex, relays)
        val transport = listOf(
            Nut18Key.Text("t") to Nut18Value.Text("nostr"),
            Nut18Key.Text("a") to Nut18Value.Text(nprofile),
            Nut18Key.Text("g") to Nut18Value.Array(
                listOf(
                    Nut18Value.Array(
                        listOf(Nut18Value.Text("n"), Nut18Value.Text(nip)),
                    ),
                ),
            ),
        )
        val request = buildList {
            add(Nut18Key.Text("i") to Nut18Value.Text(id))
            add(Nut18Key.Text("a") to amount?.takeIf { it > 0 }?.let { Nut18Value.UInt(it) }.orNull())
            add(Nut18Key.Text("u") to unit?.takeIf { it.isNotBlank() }?.let { Nut18Value.Text(it) }.orNull())
            add(Nut18Key.Text("s") to singleUse?.let { Nut18Value.Bool(it) }.orNull())
            if (mints.isNotEmpty()) add(Nut18Key.Text("m") to Nut18Value.Array(mints.map { Nut18Value.Text(it) }))
            add(Nut18Key.Text("d") to description?.takeIf { it.isNotBlank() }?.let { Nut18Value.Text(it) }.orNull())
            add(Nut18Key.Text("t") to Nut18Value.Array(listOf(Nut18Value.Map(transport))))
            // Optional NUT-10 lock (NUT-18). A payer's wallet reads this and locks
            // the proofs it creates to the given P2PK pubkey, so only its holder
            // can redeem them. Encoded as cashu-ts does: `"nut10": {"k": kind, "d": data}`.
            if (!p2pkPubkeyHex.isNullOrBlank()) {
                add(
                    Nut18Key.Text("nut10") to Nut18Value.Map(
                        listOf(
                            Nut18Key.Text("k") to Nut18Value.Text("P2PK"),
                            Nut18Key.Text("d") to Nut18Value.Text(p2pkPubkeyHex),
                        ),
                    ),
                )
            }
        }
        val cbor = Nut18Cbor.encode(Nut18Value.Map(request))
        return "creqA" + Base64.getUrlEncoder().encodeToString(cbor)
    }

    fun makeNprofile(pubkeyHex: String, relays: List<String>): String {
        val pubkey = NIP44.hexToBytes(pubkeyHex)
        require(pubkey.size == 32) { "Invalid Nostr public key." }
        val tlv = mutableListOf<Byte>()
        tlv += 0x00.toByte()
        tlv += pubkey.size.toByte()
        tlv += pubkey.toList()
        relays.forEach { relay ->
            val bytes = relay.toByteArray(Charsets.UTF_8)
            if (bytes.size <= 255) {
                tlv += 0x01.toByte()
                tlv += bytes.size.toByte()
                tlv += bytes.toList()
            }
        }
        return Bech32.encode("nprofile", tlv.toByteArray())
    }
}

/**
 * Builds the "receive locked ecash" artifact: a NUT-18 Cashu payment request that
 * locks any payment to the wallet's primary (seed-derived) P2PK key and routes the
 * proofs back over Nostr. Anyone who pays it sends ecash that only this wallet can
 * redeem. Shared by the Locked Ecash settings hub (iOS LockedReceiveRequest).
 */
object LockedReceiveRequest {
    fun build(
        nostrService: NostrService,
        settingsManager: SettingsManager,
        amount: Long? = null,
    ): String? {
        val nostrPubkey = nostrService.state.value.publicKeyHex.takeIf { it.isNotBlank() } ?: return null
        val primary = settingsManager.primaryP2PKKeyInfo() ?: return null
        val relays = settingsManager.state.value.nostrRelays.takeIf { it.isNotEmpty() } ?: return null
        return runCatching {
            PaymentRequestBuilder.build(
                id = com.cashu.me.Models.CashuRequest.newId(),
                amount = amount,
                unit = "sat",
                mints = emptyList(),
                description = null,
                nostrPubkeyHex = nostrPubkey,
                relays = relays,
                p2pkPubkeyHex = primary.publicKey,
            )
        }.getOrNull()
    }
}

private sealed interface Nut18Key {
    data class Text(val value: String) : Nut18Key
}

private sealed interface Nut18Value {
    data class Text(val value: String) : Nut18Value
    data class UInt(val value: Long) : Nut18Value
    data class Bool(val value: Boolean) : Nut18Value
    data class Array(val values: List<Nut18Value>) : Nut18Value
    data class Map(val values: List<Pair<Nut18Key, Nut18Value>>) : Nut18Value
    data object Null : Nut18Value
}

private fun Nut18Value?.orNull(): Nut18Value = this ?: Nut18Value.Null

private object Nut18Cbor {
    fun encode(value: Nut18Value): ByteArray {
        val output = mutableListOf<Byte>()
        encodeInto(value, output)
        return output.toByteArray()
    }

    private fun encodeInto(value: Nut18Value, output: MutableList<Byte>) {
        when (value) {
            is Nut18Value.Text -> {
                val bytes = value.value.toByteArray(Charsets.UTF_8)
                writeHeader(majorType = 3, length = bytes.size.toLong(), output)
                output += bytes.toList()
            }
            is Nut18Value.UInt -> writeHeader(majorType = 0, length = value.value, output)
            is Nut18Value.Bool -> output += if (value.value) 0xF5.toByte() else 0xF4.toByte()
            Nut18Value.Null -> output += 0xF6.toByte()
            is Nut18Value.Array -> {
                writeHeader(majorType = 4, length = value.values.size.toLong(), output)
                value.values.forEach { encodeInto(it, output) }
            }
            is Nut18Value.Map -> {
                writeHeader(majorType = 5, length = value.values.size.toLong(), output)
                value.values.forEach { (key, nestedValue) ->
                    when (key) {
                        is Nut18Key.Text -> encodeInto(Nut18Value.Text(key.value), output)
                    }
                    encodeInto(nestedValue, output)
                }
            }
        }
    }

    private fun writeHeader(majorType: Int, length: Long, output: MutableList<Byte>) {
        require(length >= 0) { "CBOR length cannot be negative." }
        val major = majorType shl 5
        when {
            length < 24 -> output += (major or length.toInt()).toByte()
            length < 0x100 -> {
                output += (major or 24).toByte()
                output += length.toByte()
            }
            length < 0x10000 -> {
                output += (major or 25).toByte()
                output += ((length shr 8) and 0xFF).toByte()
                output += (length and 0xFF).toByte()
            }
            length < 0x1_0000_0000L -> {
                output += (major or 26).toByte()
                for (shift in 24 downTo 0 step 8) output += ((length shr shift) and 0xFF).toByte()
            }
            else -> {
                output += (major or 27).toByte()
                for (shift in 56 downTo 0 step 8) output += ((length shr shift) and 0xFF).toByte()
            }
        }
    }
}
